src/main/java/com/meta/cp4m/message/ThreadState.java

```diff
@@ -33,20 +33,17 @@ private ThreadState(T message) {
   private ThreadState(ThreadState<T> old, T newMessage) {
     Objects.requireNonNull(newMessage);
     Preconditions.checkArgument(
-        newMessage.role() != Role.SYSTEM, "ThreadState should never hold a system message");
+            newMessage.role() != Role.SYSTEM, "ThreadState should never hold a system message");
     messageFactory = old.messageFactory;
     Preconditions.checkArgument(
         old.tail().threadId().equals(newMessage.threadId()),
         "all messages in a thread must have the same thread id");
     List<T> messages = old.messages;
-    if (newMessage.timestamp().isBefore(old.tail().timestamp())) {
-      this.messages =
-          Stream.concat(messages.stream(), Stream.of(newMessage))
-              .sorted(Comparator.comparing(Message::timestamp))
+    T mWithParentMessage = newMessage.role() == Role.USER ? (T) newMessage.addParentMessage(old.tail()): newMessage;
```

Comment:
You have an unchecked cast here, this should very rarely need to happen. Basically what this message is saying is that there is that the compiler cannot guarantee that `newMessage` and `old.tail()` will be of the same type, (which is the whole reason for using an strongly typed language 🙂). You are in a bit of a pickle here though as to get around this you'd really need a new generic on Message indicating the type of the parent message which isn't ideal. Let me think on this for a bit